### ISSUE TITLE
Dodge reduces Hit chance

### ISSUE THREAD

**Issue by Iridar** [X2CommunityCore/X2WOTCCommunityHighlander#1200]
This is especially noticeable at high Dodge values.

Firaxis made Crit be rolled into Hit, meaning that if the attack has 90% hit and 10% crit, the result chance table will look like this:

10% crit | 80% hit | 10% miss

But for Dodge they decided to implement it as a proper percentage of the overall hit chance instead. For example, if your final hit chance is 80%, and target has 50% dodge, the hit chance table you *should* get:

40% hit | 40% graze | 20% miss

But the percentage math is broken, so it doesn't work out like that.

For example, here are some values I got from testing:

Initial data: 65 Aim | 33 Crit | 90 Dodge

Final hit chance: 65

Miss chance: 35

Effective Dodge = Aim * Dodge / 100 = 65 * 90 / 100 = 59

Hit = Aim - Crit - Effective Dodge = 65 - 33 - 59 = -27

Hit result table: 

-27 Hit | 33 Crit | 59 Graze

The negative hit chance shifts the entire table to the left, so it becomes:

0 Hit | 6 Crit | 32 Graze | 62 Miss

To hammer the point home: we've gone from 65% hit and 33% crit chance to 0% hit and 6% crit chance, all because of 90 Dodge. And you won't see this in the hit chance the game shows you.

**Comment by Iridar** [comment]
On further inspection, it appears the Hit and Dodge calculation is done properly, the problem is that Crit is also subtracted from Hit. 

In other words, Dodge chance is subtracted from Hit chance, but has no effect on Crit chance, so when the math subtracts Crit from Hit, the Hit chance can get pushed well below zero, pushing the entire table to the left, and therefore reducing the chance to crit as well. 

In other words, the problem is that Hit is affected by Dodge, but Crit isn't. 

Which brings up an interesting question: should Dodge consume Crit first or Hit? Or affect both?

**Comment by Iridar** [comment]
Dodge chance replaces a percentage of Final Hit Chance with Graze hit result. 

Final Hit Chance is Hit chance + Crit chance folded into it. 

So, logically, Dodge should apply to Crit too.

However, within Final Hit Chance, crit does take priority. Meaning that it Final Hit Chance is below initial Crit chance, any hits will be crits. 

Therefore - as much as I hate it - for consistency with base game mechanics, Dodge should exhaust Hit chance first, and then Crit.

**Comment by BlackDog86** [comment]
Tested this with units of varying degrees of dodge - the output is very good.
```
Attacking unit:  (65% Aim | 33% Crit) | Defending Unit: (Dodge = 0)   Graze: 0%  | Hit: 32% | Crit: 33% | Miss: 35%
Attacking unit:  (65% Aim | 33% Crit) | Defending Unit: (Dodge = 15) Graze: 8%  | Hit: 28% | Crit: 29% | Miss: 35%
Attacking unit:  (65% Aim | 33% Crit) | Defending Unit: (Dodge = 25) Graze: 16% | Hit: 24% | Crit: 25% | Miss: 35%
Attacking unit:  (65% Aim | 33% Crit) | Defending Unit: (Dodge = 40) Graze: 25% | Hit: 20% | Crit: 20% | Miss: 35%
Attacking unit:  (65% Aim | 33% Crit) | Defending Unit: (Dodge = 50) Graze: 32% | Hit: 17% | Crit: 16% | Miss: 35%
Attacking unit:  (65% Aim | 33% Crit) | Defending Unit: (Dodge = 80) Graze: 51% | Hit: 7%  | Crit: 7%  | Miss: 35%
```
